File: src/AddTwoNumbers.java
// https://leetcode.com/problems/add-two-numbers
// T: O(max(|l1|, |l2|))
// S: O(max(|l1|, |l2|))

public class AddTwoNumbers {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode result = new ListNode();
        ListNode current = result;
        int carry = 0;

        while (l1 != null || l2 != null) {
            current.next = new ListNode();
            current = current.next;
            int sum = get(l1) + get(l2) + carry;
            current.val = sum % 10;
            carry = sum / 10;
            l1 = next(l1);
            l2 = next(l2);
        }
        if (carry > 0) {
            current.next = new ListNode(carry);
        }

        return result.next;
    }

    private ListNode next(ListNode node) {
        if (node == null) return null;
        return node.next;
    }

    private int get(ListNode node) {
        if (node == null) return 0;
        return node.val;
    }
}

File: javascript/AddTwoNumbers.js
// https://leetcode.com/problems/add-two-numbers/
// T: O(max(|l1|, |l2|))
// S: O(max(|l1|, |l2|))

const addTwoNumbers = (l1, l2) => {
  let p1 = l1;
  let p2 = l2;
  let carry = 0;
  let result_node = new ListNode(-1);
  let pointer = result_node;
  while (p1 || p2 || carry !== 0) {
    let sum = 0;
    if (p1 !== null && p2 !== null) {
      sum = p1.val + p2.val;
      p1 = p1.next;
      p2 = p2.next;
    } else if (p1 !== null) {
      sum = p1.val;
      p1 = p1.next;
    } else if (p2 !== null) {
      sum = p2.val;
      p2 = p2.next;
    }

    if (carry !== 0) {
      sum = sum + carry;
      carry = 0;
    }

    if (sum > 9) {
      carry = Math.floor(sum / 10);
      sum = sum % 10;
    }

    pointer.next = new ListNode(sum);
    pointer = pointer.next;
  }
  return result_node.next;
};
